src/D2L.CodeStyle.Analyzers/Immutability/ImmutableAttributeConsistencyChecker.cs

```diff
@@ -94,8 +94,38 @@ CancellationToken cancellationToken
 					return;
 
 				case ImmutableTypeKind.Total:
+					bool missingConditionalParameterUsage = false;
 					if( typeInfo.Kind == ImmutableTypeKind.Total ) {
-						break;
+						if( !typeInfo.IsConditional ) {
+							break;
+						}
+
+						if( baseTypeInfo.IsConditional ) {
+							foreach( ITypeParameterSymbol typeParameter in typeInfo.ConditionalTypeParameters ) {
+								bool parameterUsed = false;
+								for( int i = 0; i < baseTypeInfo.Type.TypeArguments.Length && !parameterUsed; i++ ) {
+									ITypeSymbol typeArgument = baseTypeInfo.Type.TypeArguments[ i ];
+									if( !SymbolEqualityComparer.Default.Equals( typeParameter, typeArgument ) ) {
+										continue;

```

Comment:
So this would emit a diagnostic for `IMyReadOnlyDictionary` right?

```cs
[ConditionallyImmutable]
MyKeyValuePair<[ConditionallyImmutable.ConditionallyImmutable OnlyIf]TKey, [ConditionallyImmutable.OnlyIf]TValue>

[ConditionallyImmutable]
IMyReadOnlyCollection<[ConditionallyImmutable.ConditionallyImmutable OnlyIf]T>

[ConditionallyImmutable]
IMyReadOnlyDictionary<[ConditionallyImmutable.ConditionallyImmutable OnlyIf]TKey, [ConditionallyImmutable.OnlyIf]TValue> : IMyReadOnlyCollection<KeyValuePair<TKey, TValue>>
```

because `TKey` "isn't used" in the type arguments to `IMyReadOnlyCollection`, right?

Reply:
Yeah, it would

Reply:
You thinking a recursive search?

Reply:
Ok -- you agree that the above is safe right?

Not saying its a problem with this PR, just wrapping my head around the issue

Reply:
> Ok -- you agree that the above is safe/we would like it to be accepted right?

Yep

Reply:
> You thinking a recursive search?

Something like that but I think it's a maybe a little fancy.

I think what we want to know is, are we (`IMyReadOnlyDictionary`) "at least" as immutable as all of the interfaces (base class) we implement (derive from)? "At least" meaning that any time the broader type gets judged as immutable we also do...

I think the reason its safe here is `IMyReadOnlyCollection<T>` is only immutable when T is, ok so recursively we need to care about when KeyValuePair<TKey, TValue> is... again recursively we look at that and see we have conditionals on TKey and TValue....

If we accumulate all those conditionals then I think we want to say OK, are the set of conditionals from my equal to (or a superset of?) the conditionals from my type parameters?

I'm not quite sure though, still thinking about it.